packages/diagrams/frontend/sirius-components-diagrams/src/renderer/key-binding/useDiagramKeyBinding.ts

```diff
@@ -0,0 +1,161 @@
+import { GQLPalette, GQLTool } from '../palette/Palette.types';
+import { usePaletteContents } from '../palette/usePaletteContents';
+import { useSingleClickTool } from '../tools/useSingleClickTool';
+import { UseDiagramKeyBindingState, UseDiagramKeyBindingValue } from './useDiagramKeyBinding.types';
+
+const getAllPaletteTools = (palette: GQLPalette) => {
+  return palette.quickAccessTools
+    .concat(palette.paletteEntries.filter(isTool))
+    .concat(palette.paletteEntries.filter(isToolSection).flatMap((section) => section.tools));
+};
+
+const computeCursorPosition = (
+  mousePosition: UseMousePositionValue,
+  viewport: Viewport,
+  bounds: DOMRect | undefined
+): XYPosition => {
+  const { x: mouseX, y: mouseY } = mousePosition;
+  const { x: viewportX, y: viewportY, zoom: viewportZoom } = viewport;
+  let x: number = 0;
+  let y: number = 0;
+  if (viewportZoom !== 0 && mouseX && mouseY) {
+    x = (mouseX - (bounds?.left ?? 0) - viewportX) / viewportZoom;
+    y = (mouseY - (bounds?.top ?? 0) - viewportY) / viewportZoom;
+  }
+  return { x, y };
+};
+
+const toolHasKeyBinding = (tool: GQLTool, event: React.KeyboardEvent<Element>) => {
+  return tool.keyBindings.some(
+    (keyBinding) =>
+      keyBinding.key === event.key &&
+      keyBinding.isCtrl === event.ctrlKey &&
+      keyBinding.isAlt === event.altKey &&
+      keyBinding.isMeta === event.metaKey
+  );
+};
+
+export const useDiagramKeyBinding = (diagramTargetObjectId: string): UseDiagramKeyBindingValue => {
+  const { editingContextId, diagramId } = useContext<DiagramContextValue>(DiagramContext);
+  const { getNodes, getNode, getEdges, getEdge } = useReactFlow<Node<NodeData>, Edge<EdgeData>>();
+  const [state, setState] = useState<UseDiagramKeyBindingState>({
+    currentEvent: null,
+  });
+
+  const mousePosition = useMousePosition();
+  const viewport = useViewport();
+  const store = useStoreApi<Node<NodeData>, Edge<EdgeData>>();
+  const { addErrorMessage } = useMultiToast();
+
+  const { invokeSingleClickTool } = useSingleClickTool();
+
+  const getSelectedDiagramElementIds = () => {
+    return getNodes()
+      .filter((node) => node.selected)
+      .map((node) => node.id)
+      .concat(
+        getEdges()
+          .filter((edge) => edge.selected)
+          .map((edge) => edge.id)
+      );
+  };
+
+  const getTargetObjectId = (selectedDiagramElementIds: string[]) => {
+    return (
+      getNode(selectedDiagramElementIds[0] || '')?.data.targetObjectId ||
+      getEdge(selectedDiagramElementIds[0] || '')?.data?.targetObjectId ||
+      diagramTargetObjectId
+    );
+  };
+
+  const { getPaletteContents, palette, loading } = usePaletteContents();
+
+  useEffect(() => {
+    if (!loading && palette && state.currentEvent) {
+      const allPaletteTools = getAllPaletteTools(palette);
+      const toolToExecute = allPaletteTools.filter((tool) => toolHasKeyBinding(tool, state.currentEvent!));
+      state.currentEvent.preventDefault();
```

Comment:
TODO @gdaniel this cannot work: this is not happening in the bubbling chain of the event.
`preventDefault` needs to be called in `onKeyBinding` to work